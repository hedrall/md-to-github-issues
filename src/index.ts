import { parse as _parse } from 'md-list-parser';
const input = `
    
inline 1

# heading1
## heading2

\`\`\`
    code block
\`\`\`


- (4) a
    - a-body
      - a-body-1
      - a-body-2
- (4) b
    - b-body
    - (2) b1
        - (1) b1-1
            - (b1-1-body)
        - (1) b1-2
            - b1-2-body
    - (2) b2
        - b2-body
    - b-body-2
    
    
### heading3

inline 2
`;

type Item = {
  title: string;
  paths: string[];
  depth: number;
  body: string;
  estimates: number | null;
  children: Item[];
  parents: Item[];
  isEnd: boolean;
}

const issueRegExp = /\((\d+)\).*$/;
// const transformData = (parsed: ReturnType<typeof parse>, parents: Item[] = []): Item[] => {
//   let output: Item[] = [];
//
//   parsed?.forEach(item => {
//     const {title, children} = item;
//     const paths = parents.map(p => p.title);
//     let item = {
//       title: title.replace(/\(\d+\)\s/, ''),
//       paths,
//       body: child.title,
//       estimates: Number(title.match(/\((\d+)\)/)?.[1]) || null
//     };
//
//     if (!children) return;
//
//     children.forEach(child => {
//       if (child.children) {
//         output = [...output, ...transformData([child], [...parents, item])];
//       } else {
//         output.push(item);
//       }
//     });
//   });
//
//   return output;
// };


type Structured = Exclude<ReturnType<typeof _parse>, undefined>[number];

const toMd = (parsed: Structured[], indent = 0): string => {
  return parsed.map(item => {
    const {title, children} = item;
    return [
        [' '.repeat(indent), '- ', title].join(''),
        ...(children ? [toMd(children, indent + 2)] : []),
    ].join('\n');
  }).join('\n');
};
const parseItems = (parsed: Structured[], parents: Item[] = []): Item[] => {
  return parsed.flatMap(item => {
    const { title, children } = item;
    if (!issueRegExp.test(title)) return [];
    // è¦ªNode
    const node: Item = {
      title,
      paths: parents.map(p => p.title),
      depth: parents.length + 1,
      body: children ? toMd(children) : '',
      estimates: Number(title.match(issueRegExp)?.[1]) || null,
      parents,
      children: [],
      isEnd: false,
    };
    const childNodes = parseItems(children || [], [...parents, node]);
    node.children.push(...childNodes);
    node.isEnd = childNodes.length === 0;
    return [node, ...childNodes];
  });
}

export const parse = (input: string) => {
  const _parsed = _parse(input);
  if (!_parsed) return [];

  return parseItems(_parsed);
};

export const extractIssues = (parsed: Item[]) => {
  return parsed.filter(item => item.isEnd).map(i => {
    const {title, paths, body, estimates} = i;
    return {
      title: [...paths, title].join(' > '),
      body,
      estimates,
    }
  });
}
