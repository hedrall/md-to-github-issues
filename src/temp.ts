import * as MTGI from '../src/index';

const input = `
- (8) Implement new Feature
    - (4) Database
        - Create new table
          - \`users\` table
          - \`customers\` table
    - (4) Backend
        - Create new API
        - (2) Create user API
            - (1) implementation
                - (memo ...)
            - (1) test
                - (momo memo ...)
        - (2) Create Customer API
            - (memo memo memo ...)
`;

console.table(
    MTGI.parse(input).map(i =>{
        const {body, parents, children, ...rest} = i;
        let shortBody = body.slice(0, 20);
        return {
            ...rest,
            body: shortBody + (shortBody.length === 20 ? '...' : ''),
        }
    })
)
console.log(
    MTGI.extractIssues(
        MTGI.parse(input)
    )
)
