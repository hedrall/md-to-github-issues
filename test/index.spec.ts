import { parse } from '../src/index';

test('README case', () => {
    const input = `
- (4) a
    - a-body
      - a-body-1
      - a-body-2
- (4) b
    - b-body
    - (2) b1
        - (1) b1-1
            - (b1-1-body)
        - (1) b1-2
            - b1-2-body
    - (2) b2
        - b2-body
    - b-body-2
`;
    expect(parse(input)).toMatchSnapshot();
});
