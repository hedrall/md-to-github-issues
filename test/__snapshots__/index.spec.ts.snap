// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`README case 1`] = `
[
  {
    "body": "- a-body
  - a-body-1
  - a-body-2",
    "children": [],
    "depth": 1,
    "estimates": 4,
    "isEnd": true,
    "parents": [],
    "paths": [],
    "title": "(4) a",
  },
  {
    "body": "- b-body
- (2) b1
  - (1) b1-1
    - (b1-1-body)
  - (1) b1-2
    - b1-2-body
- (2) b2
  - b2-body
- b-body-2",
    "children": [
      {
        "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
        "children": [
          {
            "body": "- (b1-1-body)",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [
              [Circular],
              [Circular],
            ],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-1",
          },
          {
            "body": "- b1-2-body",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [
              [Circular],
              [Circular],
            ],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-2",
          },
        ],
        "depth": 2,
        "estimates": 2,
        "isEnd": false,
        "parents": [
          [Circular],
        ],
        "paths": [
          "(4) b",
        ],
        "title": "(2) b1",
      },
      {
        "body": "- (b1-1-body)",
        "children": [],
        "depth": 3,
        "estimates": 1,
        "isEnd": true,
        "parents": [
          [Circular],
          {
            "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
            "children": [
              [Circular],
              {
                "body": "- b1-2-body",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-2",
              },
            ],
            "depth": 2,
            "estimates": 2,
            "isEnd": false,
            "parents": [
              [Circular],
            ],
            "paths": [
              "(4) b",
            ],
            "title": "(2) b1",
          },
        ],
        "paths": [
          "(4) b",
          "(2) b1",
        ],
        "title": "(1) b1-1",
      },
      {
        "body": "- b1-2-body",
        "children": [],
        "depth": 3,
        "estimates": 1,
        "isEnd": true,
        "parents": [
          [Circular],
          {
            "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
            "children": [
              {
                "body": "- (b1-1-body)",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-1",
              },
              [Circular],
            ],
            "depth": 2,
            "estimates": 2,
            "isEnd": false,
            "parents": [
              [Circular],
            ],
            "paths": [
              "(4) b",
            ],
            "title": "(2) b1",
          },
        ],
        "paths": [
          "(4) b",
          "(2) b1",
        ],
        "title": "(1) b1-2",
      },
      {
        "body": "- b2-body",
        "children": [],
        "depth": 2,
        "estimates": 2,
        "isEnd": true,
        "parents": [
          [Circular],
        ],
        "paths": [
          "(4) b",
        ],
        "title": "(2) b2",
      },
    ],
    "depth": 1,
    "estimates": 4,
    "isEnd": false,
    "parents": [],
    "paths": [],
    "title": "(4) b",
  },
  {
    "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
    "children": [
      {
        "body": "- (b1-1-body)",
        "children": [],
        "depth": 3,
        "estimates": 1,
        "isEnd": true,
        "parents": [
          {
            "body": "- b-body
- (2) b1
  - (1) b1-1
    - (b1-1-body)
  - (1) b1-2
    - b1-2-body
- (2) b2
  - b2-body
- b-body-2",
            "children": [
              [Circular],
              [Circular],
              {
                "body": "- b1-2-body",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-2",
              },
              {
                "body": "- b2-body",
                "children": [],
                "depth": 2,
                "estimates": 2,
                "isEnd": true,
                "parents": [
                  [Circular],
                ],
                "paths": [
                  "(4) b",
                ],
                "title": "(2) b2",
              },
            ],
            "depth": 1,
            "estimates": 4,
            "isEnd": false,
            "parents": [],
            "paths": [],
            "title": "(4) b",
          },
          [Circular],
        ],
        "paths": [
          "(4) b",
          "(2) b1",
        ],
        "title": "(1) b1-1",
      },
      {
        "body": "- b1-2-body",
        "children": [],
        "depth": 3,
        "estimates": 1,
        "isEnd": true,
        "parents": [
          {
            "body": "- b-body
- (2) b1
  - (1) b1-1
    - (b1-1-body)
  - (1) b1-2
    - b1-2-body
- (2) b2
  - b2-body
- b-body-2",
            "children": [
              [Circular],
              {
                "body": "- (b1-1-body)",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-1",
              },
              [Circular],
              {
                "body": "- b2-body",
                "children": [],
                "depth": 2,
                "estimates": 2,
                "isEnd": true,
                "parents": [
                  [Circular],
                ],
                "paths": [
                  "(4) b",
                ],
                "title": "(2) b2",
              },
            ],
            "depth": 1,
            "estimates": 4,
            "isEnd": false,
            "parents": [],
            "paths": [],
            "title": "(4) b",
          },
          [Circular],
        ],
        "paths": [
          "(4) b",
          "(2) b1",
        ],
        "title": "(1) b1-2",
      },
    ],
    "depth": 2,
    "estimates": 2,
    "isEnd": false,
    "parents": [
      {
        "body": "- b-body
- (2) b1
  - (1) b1-1
    - (b1-1-body)
  - (1) b1-2
    - b1-2-body
- (2) b2
  - b2-body
- b-body-2",
        "children": [
          [Circular],
          {
            "body": "- (b1-1-body)",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [
              [Circular],
              [Circular],
            ],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-1",
          },
          {
            "body": "- b1-2-body",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [
              [Circular],
              [Circular],
            ],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-2",
          },
          {
            "body": "- b2-body",
            "children": [],
            "depth": 2,
            "estimates": 2,
            "isEnd": true,
            "parents": [Circular],
            "paths": [
              "(4) b",
            ],
            "title": "(2) b2",
          },
        ],
        "depth": 1,
        "estimates": 4,
        "isEnd": false,
        "parents": [],
        "paths": [],
        "title": "(4) b",
      },
    ],
    "paths": [
      "(4) b",
    ],
    "title": "(2) b1",
  },
  {
    "body": "- (b1-1-body)",
    "children": [],
    "depth": 3,
    "estimates": 1,
    "isEnd": true,
    "parents": [
      {
        "body": "- b-body
- (2) b1
  - (1) b1-1
    - (b1-1-body)
  - (1) b1-2
    - b1-2-body
- (2) b2
  - b2-body
- b-body-2",
        "children": [
          {
            "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
            "children": [
              [Circular],
              {
                "body": "- b1-2-body",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-2",
              },
            ],
            "depth": 2,
            "estimates": 2,
            "isEnd": false,
            "parents": [
              [Circular],
            ],
            "paths": [
              "(4) b",
            ],
            "title": "(2) b1",
          },
          [Circular],
          {
            "body": "- b1-2-body",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [Circular],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-2",
          },
          {
            "body": "- b2-body",
            "children": [],
            "depth": 2,
            "estimates": 2,
            "isEnd": true,
            "parents": [
              [Circular],
            ],
            "paths": [
              "(4) b",
            ],
            "title": "(2) b2",
          },
        ],
        "depth": 1,
        "estimates": 4,
        "isEnd": false,
        "parents": [],
        "paths": [],
        "title": "(4) b",
      },
      {
        "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
        "children": [
          [Circular],
          {
            "body": "- b1-2-body",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [Circular],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-2",
          },
        ],
        "depth": 2,
        "estimates": 2,
        "isEnd": false,
        "parents": [
          {
            "body": "- b-body
- (2) b1
  - (1) b1-1
    - (b1-1-body)
  - (1) b1-2
    - b1-2-body
- (2) b2
  - b2-body
- b-body-2",
            "children": [
              [Circular],
              [Circular],
              {
                "body": "- b1-2-body",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-2",
              },
              {
                "body": "- b2-body",
                "children": [],
                "depth": 2,
                "estimates": 2,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                ],
                "title": "(2) b2",
              },
            ],
            "depth": 1,
            "estimates": 4,
            "isEnd": false,
            "parents": [],
            "paths": [],
            "title": "(4) b",
          },
        ],
        "paths": [
          "(4) b",
        ],
        "title": "(2) b1",
      },
    ],
    "paths": [
      "(4) b",
      "(2) b1",
    ],
    "title": "(1) b1-1",
  },
  {
    "body": "- b1-2-body",
    "children": [],
    "depth": 3,
    "estimates": 1,
    "isEnd": true,
    "parents": [
      {
        "body": "- b-body
- (2) b1
  - (1) b1-1
    - (b1-1-body)
  - (1) b1-2
    - b1-2-body
- (2) b2
  - b2-body
- b-body-2",
        "children": [
          {
            "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
            "children": [
              {
                "body": "- (b1-1-body)",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-1",
              },
              [Circular],
            ],
            "depth": 2,
            "estimates": 2,
            "isEnd": false,
            "parents": [
              [Circular],
            ],
            "paths": [
              "(4) b",
            ],
            "title": "(2) b1",
          },
          {
            "body": "- (b1-1-body)",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [Circular],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-1",
          },
          [Circular],
          {
            "body": "- b2-body",
            "children": [],
            "depth": 2,
            "estimates": 2,
            "isEnd": true,
            "parents": [
              [Circular],
            ],
            "paths": [
              "(4) b",
            ],
            "title": "(2) b2",
          },
        ],
        "depth": 1,
        "estimates": 4,
        "isEnd": false,
        "parents": [],
        "paths": [],
        "title": "(4) b",
      },
      {
        "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
        "children": [
          {
            "body": "- (b1-1-body)",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [Circular],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-1",
          },
          [Circular],
        ],
        "depth": 2,
        "estimates": 2,
        "isEnd": false,
        "parents": [
          {
            "body": "- b-body
- (2) b1
  - (1) b1-1
    - (b1-1-body)
  - (1) b1-2
    - b1-2-body
- (2) b2
  - b2-body
- b-body-2",
            "children": [
              [Circular],
              {
                "body": "- (b1-1-body)",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-1",
              },
              [Circular],
              {
                "body": "- b2-body",
                "children": [],
                "depth": 2,
                "estimates": 2,
                "isEnd": true,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                ],
                "title": "(2) b2",
              },
            ],
            "depth": 1,
            "estimates": 4,
            "isEnd": false,
            "parents": [],
            "paths": [],
            "title": "(4) b",
          },
        ],
        "paths": [
          "(4) b",
        ],
        "title": "(2) b1",
      },
    ],
    "paths": [
      "(4) b",
      "(2) b1",
    ],
    "title": "(1) b1-2",
  },
  {
    "body": "- b2-body",
    "children": [],
    "depth": 2,
    "estimates": 2,
    "isEnd": true,
    "parents": [
      {
        "body": "- b-body
- (2) b1
  - (1) b1-1
    - (b1-1-body)
  - (1) b1-2
    - b1-2-body
- (2) b2
  - b2-body
- b-body-2",
        "children": [
          {
            "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
            "children": [
              {
                "body": "- (b1-1-body)",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [
                  [Circular],
                  [Circular],
                ],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-1",
              },
              {
                "body": "- b1-2-body",
                "children": [],
                "depth": 3,
                "estimates": 1,
                "isEnd": true,
                "parents": [
                  [Circular],
                  [Circular],
                ],
                "paths": [
                  "(4) b",
                  "(2) b1",
                ],
                "title": "(1) b1-2",
              },
            ],
            "depth": 2,
            "estimates": 2,
            "isEnd": false,
            "parents": [Circular],
            "paths": [
              "(4) b",
            ],
            "title": "(2) b1",
          },
          {
            "body": "- (b1-1-body)",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [
              [Circular],
              {
                "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
                "children": [
                  [Circular],
                  {
                    "body": "- b1-2-body",
                    "children": [],
                    "depth": 3,
                    "estimates": 1,
                    "isEnd": true,
                    "parents": [Circular],
                    "paths": [
                      "(4) b",
                      "(2) b1",
                    ],
                    "title": "(1) b1-2",
                  },
                ],
                "depth": 2,
                "estimates": 2,
                "isEnd": false,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                ],
                "title": "(2) b1",
              },
            ],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-1",
          },
          {
            "body": "- b1-2-body",
            "children": [],
            "depth": 3,
            "estimates": 1,
            "isEnd": true,
            "parents": [
              [Circular],
              {
                "body": "- (1) b1-1
  - (b1-1-body)
- (1) b1-2
  - b1-2-body",
                "children": [
                  {
                    "body": "- (b1-1-body)",
                    "children": [],
                    "depth": 3,
                    "estimates": 1,
                    "isEnd": true,
                    "parents": [Circular],
                    "paths": [
                      "(4) b",
                      "(2) b1",
                    ],
                    "title": "(1) b1-1",
                  },
                  [Circular],
                ],
                "depth": 2,
                "estimates": 2,
                "isEnd": false,
                "parents": [Circular],
                "paths": [
                  "(4) b",
                ],
                "title": "(2) b1",
              },
            ],
            "paths": [
              "(4) b",
              "(2) b1",
            ],
            "title": "(1) b1-2",
          },
          [Circular],
        ],
        "depth": 1,
        "estimates": 4,
        "isEnd": false,
        "parents": [],
        "paths": [],
        "title": "(4) b",
      },
    ],
    "paths": [
      "(4) b",
    ],
    "title": "(2) b2",
  },
]
`;
